presto-main/src/main/java/com/facebook/presto/metadata/MetadataManager.java

```diff
@@ -430,6 +431,18 @@ public TableLayoutHandle getAlternativeLayoutHandle(Session session, TableLayout
         return commonHandle.map(handle -> new PartitioningHandle(Optional.of(connectorId), left.getTransactionHandle(), handle));
     }
 
+    @Override
+    public PartitioningHandle getPartitioningHandleForExchange(Session session, String catalogName, int partitionCount, List<Type> partitionTypes)
+    {
+        CatalogMetadata catalogMetadata = getOptionalCatalogMetadata(session, catalogName)
+                .orElseThrow(() -> new PrestoException(NOT_FOUND, format("Catalog '%s' does not exist", catalogName)));
```

Comment:
Should this be an `IllegalStateException`? This method seems only applicable to DML, which I think if catalog is missing, it should be caught earlier than here (not entirely sure about this).

Reply:
We should keep the exception as user error. `IllegalStateException` will be categorized as in internal error.